cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ProgramRunDispatcherTask.java

```diff
@@ -0,0 +1,152 @@
+
+  private final Injector injector;
+
+  @Inject
+  ProgramRunDispatcherTask(Injector injector) {
+    this.injector = injector;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    try {
+      ProgramRunInfo programRunInfo = GSON.fromJson(context.getParam(), ProgramRunInfo.class);
+      DispatchTaskRunner taskRunner = injector.getInstance(DispatchTaskRunner.class);
+      ProgramController programController = taskRunner.dispatch(programRunInfo);
+      RemoteProgramRunDispatcherResponse response = null;
+      if (Objects.nonNull(programController)) {
+        response = new RemoteProgramRunDispatcherResponse(true);
+      }
+      String result = GSON.toJson(response);
+      context.writeResult(result.getBytes(StandardCharsets.UTF_8));
+    } catch (Exception e) {
+      LOG.error("Exception: ", e);
+    }
+  }
+
+  private static class DispatchTaskRunner {
+
+    private final CConfiguration cConf;
+    private final ProgramRunnerFactory programRunnerFactory;
+    private final ConfiguratorFactory configuratorFactory;
+    private final Impersonator impersonator;
+    private final ArtifactRepository artifactRepository;
+    private ProgramRunnerFactory remoteProgramRunnerFactory;
+    private InetAddress host;
+
+    @Inject
+    DispatchTaskRunner(CConfiguration cConf,
+        ProgramRunnerFactory programRunnerFactory,
+        ConfiguratorFactory configuratorFactory,
+        Impersonator impersonator,
+        @Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO) ArtifactRepository artifactRepository) {
+      this.cConf = cConf;
+      this.programRunnerFactory = programRunnerFactory;
+      this.configuratorFactory = configuratorFactory;
+      this.impersonator = impersonator;
+      this.artifactRepository = artifactRepository;
+    }
+
+    /**
+     * Optional guice injection for the {@link ProgramRunnerFactory} used for remote execution. It
+     * is optional because in unit-test we don't have need for that.
+     */
+    @Inject(optional = true)
+    void setRemoteProgramRunnerFactory(
+        @Constants.AppFabric.RemoteExecution ProgramRunnerFactory runnerFactory) {
+      this.remoteProgramRunnerFactory = runnerFactory;
+    }
+
+    @Inject(optional = true)
+    void setHostname(@Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress host) {
+      this.host = host;
+    }
+
+    public ProgramController dispatch(ProgramRunInfo programRunInfo) throws Exception {
+      InMemoryProgramRunDispatcher dispatcher = new InMemoryProgramRunDispatcher(cConf,
+          programRunnerFactory, configuratorFactory, impersonator, artifactRepository);
+      dispatcher.setRemoteProgramRunnerFactory(remoteProgramRunnerFactory);
+      dispatcher.setHostname(host);
+      try {
+        return dispatcher.dispatchProgram(programRunInfo);
+      } catch (Exception e) {
+        // We don't need the ExecutionException being reported back to the RemoteTaskExecutor, hence only
```

Comment:
does this throw an ExecutionException? I thought the Dispatcher API just throws Exception? If it's not an ExecutionException, throwing the cause will lose part of the stack trace.

Even if it does throw ExecutionException, it would be cleaner to catch ExecutionException so that it's guaranteed to be an ExecutionException.

Reply:
Changed